transports/dns/src/lib.rs

```diff
@@ -85,6 +86,30 @@ pub mod async_std {
                 resolver: async_std_resolver::resolver(cfg, opts).await,
             }
         }
+
+        // TODO: Replace `system` implementation with this
+        #[doc(hidden)]
```

Comment:
To actually be able to replace this, we also need to patch `async_std_resolver` to remove the `async` from the function definition.

Reply:
I was wondering if `AsyncStdResolver::resolver_from_system_conf` was lazy initialized to avoid reading the `/etc/resolv.conf` in a non-async context, and it turns out it is reading the file on construction([1], [2]). So that call can block the executor. `async_std_resolver` needs much more changes to remove the `async`.

[1]: https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/resolver/src/async_resolver.rs#L242
[2]: https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/resolver/src/system_conf/unix.rs#L30-L39

Reply:
> This is a nice idea! So we have this here and thus consider it to be not public API, right?

That is correct.

Reply:
> I was wondering if `AsyncStdResolver::resolver_from_system_conf` was lazy initialized to avoid reading the `/etc/resolv.conf` in a non-async context, and it turns out it is reading the file on construction([1](https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/resolver/src/async_resolver.rs#L242), [2](https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/resolver/src/system_conf/unix.rs#L30-L39)). So that call can block the executor. `async_std_resolver` needs much more changes to remove the `async`.

Correct me if I am wrong but the code you linked doesn't use async? And the file is usually super small so I guess that is why they didn't bother reading it asynchronously?

Reply:
> Correct me if I am wrong but the code you linked doesn't use async? 

Yes in the deepest layer (i.e. `read_resolv_conf`) it doesn't use async and in a higher level they don't use `spawn_blocking`. This is the full chain: [1] -> [2] -> [3] -> [4].

[1]: https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/async-std-resolver/src/lib.rs#L141-L143
[2]: https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/resolver/src/async_resolver.rs#L170-L172
[3]: https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/resolver/src/async_resolver.rs#L241-L243
[4]: https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/resolver/src/system_conf/unix.rs#L30-L39

> And the file is usually super small so I guess that is why they didn't bother reading it asynchronously?

I guess yes, however someone can boot its system from an NFS. In that case the read can be very slow and it will block the executor. Pragmatically, even with NFS, this may never happen, but it is still a possibility.

Reply:
> > Correct me if I am wrong but the code you linked doesn't use async?
> 
> Yes in the deepest layer (i.e. `read_resolv_conf`) it doesn't use async and in a higher level they don't use `spawn_blocking` or another thread. This is the full chain: [1](https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/async-std-resolver/src/lib.rs#L141-L143) -> [2](https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/resolver/src/async_resolver.rs#L170-L172) -> [3](https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/resolver/src/async_resolver.rs#L241-L243) -> [4](https://github.com/hickory-dns/hickory-dns/blob/d07455476cfba08a661c09f303af2cfbb8c1a5cc/crates/resolver/src/system_conf/unix.rs#L30-L39).

Right so essentially, at the moment we have the worst of both worlds: Having to deal with the `Future` but the underlying IO does not yield when it would block :(

I'll raise an issue with them to discuss it.

Reply:
See https://github.com/hickory-dns/hickory-dns/issues/2089.